pkg/engine/diff.go

```diff
@@ -232,25 +232,32 @@ func PrintObject(
 	}
 }
 
-func massageStackPreviewAdd(p resource.PropertyValue) {
+func massageStackPreviewAdd(p resource.PropertyValue) resource.PropertyValue {
 	switch {
 	case p.IsArray():
-		for _, v := range p.ArrayValue() {
-			massageStackPreviewAdd(v)
+		arr := make([]resource.PropertyValue, len(p.ArrayValue()))
+		for i, v := range p.ArrayValue() {
+			arr[i] = massageStackPreviewAdd(v)
 		}
+		return resource.NewArrayProperty(arr)
 	case p.IsObject():
-		delete(p.ObjectValue(), "@isPulumiResource")
-		for _, v := range p.ObjectValue() {
-			massageStackPreviewAdd(v)
+		obj := resource.PropertyMap{}
+		for k, v := range p.ObjectValue() {
+			if k != "@isPulumiResource" {
```

Comment:
Nit: Another comment as for why we ignore `"@isPulumiResource"` would be helpful here. Since it isn't a constant referenced from another package, I assume this is only hard-coded in a couple of places?